drivers/ieee802154/ieee802154_dw1000.c

```diff
@@ -769,12 +769,66 @@ static int dwt_filter(const struct device *dev,
 	return -ENOTSUP;
 }
 
+// From 7.2.31.1 from User Manual
+#define DWT_TX_POWER_CTRL_OFF 7
+#define DWT_TX_POWER_CTRL_COARSE_MASK 0xE0U
+#define DWT_TX_POWER_CTRL_COARSE_STEPS 2.5
+#define DWT_TX_POWER_CTRL_FINE_MASK 0x1FU
+#define DWT_TX_POWER_CTRL_FINE_STEPS 0.5
+
+#define DWT_TX_POWER_TXPOWPHR_SHIFT 8
+#define DWT_TX_POWER_TXPOWSD_SHIFT 16
+
+// These should go into DT probably
+#define DWT_TX_POWER_BOOSTP500_MOD 3
+#define DWT_TX_POWER_BOOSTP250_MOD 6
+#define DWT_TX_POWER_BOOSTP125_MOD 9
+#define DWT_TX_POWER_PHR_MOD 0
+
+uint8_t dwt_encode_power_tx(float power) {
+	uint8_t coarse = DWT_TX_POWER_CTRL_OFF;
+	uint8_t fineTuneVal = 0;
+	if (power >= 0) {
+		coarse--;
+		while (power >= DWT_TX_POWER_CTRL_COARSE_STEPS && coarse > 0) {
```

Comment:
Is this comment still valid?

Reply:
Yes, this should still be valid - see the summary of our research below which largely confirms the assumptions made in this comment.

Reply:
```js
function calc_tx_power(target_power) {
    coarse_power_step_x2 = 5;
    fine_power_step_x2 = 1;
    max_coarse_power_step = 0b110;
    max_fine_power_step = 0b11111;

    target_power_int = Math.trunc(target_power);
    target_power_int_x2 = target_power_int << 1;

    coarse_power_steps = Math.min(div_c(target_power_int_x2, coarse_power_step_x2), max_coarse_power_step);
    coarse_power_x2 = coarse_power_steps * coarse_power_step_x2;
    fine_power_steps = div_c(target_power_int_x2 - coarse_power_x2, fine_power_step_x2);
    
    coarse_power_reg = max_coarse_power_step - coarse_power_steps;
    fine_power_reg = fine_power_steps;
    power_reg = coarse_power_reg << 5 | fine_power_reg;
    
    calc_power = coarse_power_steps * 2.5 + fine_power_steps * 0.5;
    
    console.log(`coarse pwr stps: ${coarse_power_steps}, fine pwr stps: ${fine_power_steps}, is valid: ${target_power_int === calc_power}, reg: ${power_reg.toString(16)}`);
}
```

Did it in JS so that you can play around with it in your browser console first. Of course this is a prime example of a function that deserves to be unit tested. I just did a few numbers - so no guarantee the algo is 100% correct.

Reply:
Your solution is definitely better. I've tested it a bit more and I haven't found any problem.
It's been adopted